File: starter-files/Javascript/part-01/task-02-add-error-handling/blueprint_decoder.js
const fs = require("fs");

/**
 * Reads a blueprint file and extracts all secrets marked between {* and *}
 * Example: League transmission contains {* EMERGENCY_PROTOCOL: NIGHTFALL_SEQUENCE_ACTIVE *}
 * Extracts: "EMERGENCY_PROTOCOL: NIGHTFALL_SEQUENCE_ACTIVE" (without markers)
 * Uses regex pattern with matchAll()
 * @param {string} filename - Path to the blueprint file
 * @returns {Array<string>} Array of extracted secret strings
 */
function decodeBlueprint(filename) {
    const content = fs.readFileSync(filename, "utf8");

    // Use regex to find all secrets between {* and *}
    const pattern = /\{\* (.*?) \*\}/g;
    const matches = [...content.matchAll(pattern)];
    const secrets = matches.map((match) => match[1]);

    return secrets;
}

// Enhanced version with error handling
// If file doesn't exist, return empty array and print error message
function decodeBlueprintSafe(filename) {
    // TODO: Add try-catch to handle file errors
    // TODO: If file doesn't exist, print error and return empty array
    // TODO: Otherwise, call decodeBlueprint and return secrets
    return [];
}

// Test error handling
const secrets1 = decodeBlueprintSafe("nonexistent.txt");
console.log(`Found ${secrets1.length} secrets`);

// Test normal operation
const secrets2 = decodeBlueprintSafe("blueprint-data.txt");
console.log(`Found ${secrets2.length} secrets`);

module.exports = {
    decodeBlueprint,
    decodeBlueprintSafe,
};

File: starter-files/Python/part-01/task-02-add-error-handling/blueprint_decoder.py
import re

# Function to read a blueprint file and extract all secrets marked between {* and *}
# Example: League Blueprint contains {* AGENT_CODENAME: SHADOWMIND *}
# Should extract: "AGENT_CODENAME: SHADOWMIND" (without the markers)
# Uses regex pattern to find all occurrences
# Returns a list of extracted secrets
def decode_blueprint(filename):
    with open(filename, "r") as file:
        content = file.read()

    # Use regex to find all secrets between {* and *}
    pattern = r"\{\* (.*?) \*\}"
    secrets = re.findall(pattern, content)
    return secrets


# Enhanced version with error handling
# If file doesn't exist, return an empty list and print an error message
def decode_blueprint_safe(filename):
    # TODO: Add try-except to catch FileNotFoundError
    # TODO: If file doesn't exist, print error and return empty list
    # TODO: Otherwise, call decode_blueprint and return the secrets
    pass


if __name__ == "__main__":
    # Test error handling
    secrets = decode_blueprint_safe("nonexistent.txt")
    print(f"Found {len(secrets)} secrets")

    # Test normal operation
    secrets = decode_blueprint_safe("blueprint-data.txt")
    print(f"Found {len(secrets)} secrets")
